Support custom devtools server commands

ventuno opened the pull request:
ChromeDriver version `2.30` [1] will support 2 new endpoints to send custom commands to the DevTools debugger server:

* `/chromium/send_command`;
* `/chromium/send_command_and_get_result`.

In this PR, I add the support for these new commands to the `webdriver-js-extender`.

[1] https://chromium.googlesource.com/chromium/src/+/711de0efbb675bd2a4a28ec47c9194d8e498e600

ventuno commented:
@qiyigg (or @sjelin) I figured out what was wrong with my previous PR [1]. I fixed the commits with the correct username/email address. This should be good to go.

[1] https://github.com/angular/webdriver-js-extender/pull/13

qiyigg commented:
Sorry, I am not the maintainer of webdriver_js_extender and I don't have knowledge of it. Probably @NickTomlin could review it?

ventuno commented:
Thanks @qiyigg. @NickTomlin would you mind taking a look at this?

NickTomlin commented on lib/index.ts:
```diff
@@ -141,6 +141,10 @@ export interface ExtendedWebDriver extends WebDriver {
   // See https://github.com/webdriverio/webdriverio/blob/v4.6.1/lib/protocol/shake
   shakeDevice: () => wdpromise.Promise<void>;
+
+  sendCommand: (cmd: string, params: Object) => wdpromise.Promise<void>;
```
We don't have a consistent pattern for naming, but what about something like `sendChromiumCommand` to make the connection to the chromium tools clearer?

ventuno replied:
Done.

NickTomlin commented on lib/command_definitions.ts:
```diff
@@ -146,3 +146,7 @@ export let rotationGesture = new CommandDefinition(
+export let sendCommand =
+    new CommandDefinition<void>('sendCommand', ['cmd', 'params'], 'POST', '/chromium/send_command');
+export let sendCommandAndGetResult = new CommandDefinition<Object>(
+  'sendCommandAndGetResult', ['cmd', 'params'], 'POST', '/chromium/send_command_and_get_result');
```
Just curious — since I am unfamiliar with these chromium commands — is there a practical difference between them besides not returning a result?

ventuno replied:
Not really. I'm just exposing the two commands available in the devtools client. Not all devtools commands will necessarily return a result.

NickTomlin commented:
@ventuno this is probably a bug in the existing tests. I will try to take a look at this and fix it on master first. 

I can't promise i'll get to it before this weekend unfortunately.

ventuno commented:
Thanks a lot for the feedback @NickTomlin. I didn't realize I broke some existing tests by adding the two new commands. I fixed the issue and added the new tests, see the output of running `npm test` below:

```
> gulp test

[21:31:42] Using gulpfile ~/Documents/code/webdriver-js-extender/gulpfile.js
[21:31:42] Starting 'prepublish'...
[21:31:42] Starting 'format'...
[21:31:42] Finished 'format' after 99 ms
[21:31:42] Starting 'tsc'...
[21:31:45] Finished 'tsc' after 3 s
[21:31:45] Starting 'copy'...
[21:31:46] Finished 'copy' after 480 ms
[21:31:46] Finished 'prepublish' after 3.58 s
[21:31:46] Starting 'build'...
[21:31:46] Finished 'build' after 2.67 μs
[21:31:46] Starting 'test'...
Started
.................................................................


65 specs, 0 failures
Finished in 2.919 seconds

[21:31:49] Finished 'test' after 3.35 s
```

In order to structure the new mocks, I used the same structure used for appium's custom commands. Let me know if that makes sense to you.

ventuno commented:
Thanks a lot @NickTomlin. @qiyigg how do I get this merged? Also, do you happen to know if/when this is going to be released?

qiyigg commented:
Will publish it today.

ventuno commented:
Thanks @qiyigg!
